### ISSUE TITLE
Unwanted space before DU cases when pattern matching

### ISSUE THREAD

**Issue by drhumlen** [fsprojects/fantomas#2369]
Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAXAngB0WAKjngLwA6AdhBAD4GYAyAhgLZiwBmduYAVF4hSphahAGqMATm04BndBICW5AOaCqIzAGFYlDmDmKVvfmAA2CuWoiwAbogmKkYdAAsLAOnywAyvKXKACgBKMDJKISpmRnRoF2c3GTAAdwVXKwiNJmYA9ABGABpnACYQgFoAPn1sQ3ROUkgAgFIAeTAKsBag%2Budc4vShDXEpRjbKxn71Om1yAJlC9BkyyplIfPAoGUR0dH8VtABtdYhQCKgAEixcVCgldFWjqkgzm0ZTAFdkA4ehAAY104gkAAkuQkOR0N4FAAvZD-AEAFm%2BYAAug8AL5w45Ip6XT43cH3U5PF7vPGHAFgApIwEAWUYAA96EpEPRECpXISAbkij8kaihBiHicIjiiNdIGz4M12EzyIhfJhTLDsc9Xh9ruSAUVMSKAKKg6Wy5UUyCmdiQPnonXCoSiq5oSAAI1gsFMnNtqtJGupYAAzDrbSDNhJ0AAxJSvAByiCS5jl7oi8g%2BloF1pVuPFztdCcBnvVXwp8IDj282EY0EQACFEOxYBJEAAFSQsLb2HNUJOIFNUQVCG2PC5ih1Zt3F4lqsk%2BgCsxcBpfLVZrdZZsCS9mgjE2IJssA3Ox07Ygne7EF7VH7uYzw5do5VJPzYE1pwAbLOoPOK9Xa-WAKrYXASBuW7kDue4KAeb7sK8mwnmAZ5YkSg72lAI7tuOXoFgCADsb6QB%2Bi7fogmimJuMjTAYbzQOgdaHmAUGmDBpz8j2aaIVeKE3mheaThSAAcuH4V%2By40ogzCOm2kHQV2TFWkK6ZDhx2Zjtx3oUgAnAJZafku9baKYEE%2BvRjERMxp6sSKSF4k6nHKfePFcn8Pp4VpiAAILsOg9jaMwUS0ceMmpnJbEKdZSl3hOqlclSJqCTp8qiQocD6eQtFGdJJmyX28nIaFt6IXZkWnNymkLu5nkSN4CVJQZFL%2BRlgVZcFOWobZEWYUV-pOfhrkSLAbyggAIog5jMKkEk%2BnVQimXB5keuxkC3FxBXtREuRFk5dL0kC7D4C4bLeC4dboAA6go8Acm%2BvIBSxQUWfNi2tRhj4%2BrkM4bQy227WyuoMfKh0hqd50uLRz5XfVN2NXdIUPeFT1Pqtr7vVt5DsAo9LNAB0R1rq9LVIgMgyOBKVvrxYNTZl57ZVZMP5W1z0UrkOFIwASogcASPAgMXT68Jk1Q03wWAF7nPdBKPQ%2B8NCLk-Es2zdbwJGbxifY0pAp5zArG%2BuSwYLwt2lZzBvKYOxxogoZ1lE2z2IQVzi-ZRUaU5rPs-ANJGybzLmxIlvlbRMAuM2VH2AA%2BikQMWtdZm3XN0Ni7DEs%2BjyuGbT1EiMJgzQSEychc8DJN8xAAuzQOot3HbhUREU0UApAKcOOnmfZ%2BgivKxIqvq5rL068Xl4hYbxsKKbXs%2B9bYrlytQhFNqTmpw3WcWOgbsD0PFvRL7uGxIH5Wh2dHLd9HJex2X8f25XnUmpt4ikpWSjwP4ue0aT%2B%2BQzHOU0xZy301q60Xwyob9VRImN9QT313nnHmBcUQUwQlDN%2BcdaZw0Tm9X%2B9IBqwHQAAcS2DjPGBMiYP3zs-SmTUrItRPhXSeiML7u0HsySs%2BloAAGtKxpyYVsGQzRyDeBbHpJWxNDJSSITA1%2BpCbLkInlQIoTMTTRljHQrYa42Q20QENVG5BUhE1cqCESrdO61QkMmSOM0D692amIhBCcKRFBliaVy5hlDkH-uQQBOhIQOOiG8esXgQRgnQNEImqVBFGN1lTTM5iP500lpIx2tjjb2HIP4uw9AdDKB0eJCQqilAaJ0HogEaUhFC1CdeMKFjT5CF9I5ahA8jRZEdPARgxFYCExULI02gTfoFL1pZMJJSImIIpL6auRIcH2AUMwNkfjTAAGlECIGwD4iZIIWGMGceAik%2BTgk9xFiFMhpSKFUF9NPE09CVmMKNDIHqfVQTRjkIgV2NDTY4PrHgnJfkDHpXJg1YhsDRG9I9J-KJEBfTnxrpWSQQl6wDQsNARQo0EmeXgD%2BdROghrQBImnfc-D1lBPBlHF%2Bh8zF-IHACn0vof41xGbCiZrxfC9TeBRbAColTtOMp8iG3yRHinfv8yJpLkE1xmXMzaLd0nShOeQM5zJclFR%2BJAouJjtmEryn0yxAJfRUJrrSxKi8EDGjyTitlp51jIjhAtGQoZCbXHyWiIAA)

#### Code

```fsharp
type Type =
    | TyLam of Type * Type
    | TyVar of string
    | TyCon of string * Type list
    override this.ToString() =
        match this with
        | TyLam(t1, t2) -> sprintf "(%O -> %O)" t1 t2
        | TyVar a -> a
        | TyCon(s, ts) -> s
```
            

#### Result

```fsharp
type Type =
    | TyLam of Type * Type
    | TyVar of string
    | TyCon of string * Type list

    override this.ToString() =
        match this with
        | TyLam (t1, t2) -> sprintf "(%O -> %O)" t1 t2
        | TyVar a -> a
        | TyCon (s, ts) -> s

```
            
#### Problem description

Fantomas adds a space before the parenthesis for the Discriminated Union cases when pattern matching. According to the F# style guide, there should not be such a space: https://docs.microsoft.com/en-us/dotnet/fsharp/style-guide/formatting#formatting-pattern-matching

It also makes sense to not have it since you contruct your object like `let myShape = Rect(4.0, 6.0)`, so the destructuring should look similar `match myShape with Rect(w, h) -> .....`

Also worth noting here is that the output from Fantomas does not match the example given in the README because of this behavior (see example above).

#### Extra information

- [ ] The formatted result breaks my code.
- [ ] The formatted result gives compiler warnings.
- [x] I or my company would be willing to help fix this.

#### Options

Fantomas master branch at 2022-07-15T14:30:45Z - 87cd2120c82ce2c83838ee0eaa81f53d03ed33af

Default Fantomas configuration

**Comment by nojaf** [comment]
Hello, thank you for raising this issue. It might be me, and I haven't finished my first coffee on a Monday morning, but where does the guide mention this explicitly of the space?

It makes sense to change this but this can be a bit of an impactful change for many users, so I really would like to have crystal clear guidance to be able to justify this change. A code sample in the F# style guide without any explicit text isn't really reliable.

Currently, the setting `fsharp_space_before_parameter` is being used to determine if there should be a space or not. I believe it might make more sense to use the same rules as for function invocation.
There, depending on the capitalisation of the function name, `fsharp_space_before_lowercase_invocation` or `fsharp_space_before_uppercase_invocation` is used.

**Comment by drhumlen** [comment]
> Hello, thank you for raising this issue. It might be me, and I haven't finished my first coffee on a Monday morning, but where does the guide mention this explicitly of the space?

Hmm on reading I can't find anything explicit actually, but all the examples are without the space. I also think of it as exactly the same the construction of something – ie `let shape = Rect(2.0, 6.0)`, and not `let shape = Rect (2.0, 6.0)`, so I think of it as "covered" by that rule. But some official statement on this would be nice.

> Currently, the setting fsharp_space_before_parameter is being used to determine if there should be a space or not. I believe it might make more sense to use the same rules as for function invocation.
There, depending on the capitalisation of the function name, fsharp_space_before_lowercase_invocation or fsharp_space_before_uppercase_invocation is used.

Yea I think that makes sense. It should always follow the same spacing rule as the construction of DU – which is very similar to an uppercase (invocation).

@nojaf Is there anyone the Microsoft/F# committee we can confer with? Philip Carter?

**Comment by nojaf** [comment]
Please raise this over at https://github.com/fsharp/fslang-design#style-guide.

**Comment by auduchinok** [comment]
The adding of space is very inconsistent to the expression handling. I'd expect Fantomas to use `SpaceBeforeUppercaseInvocation` setting here.

**Comment by auduchinok** [comment]
Created an issue: https://github.com/fsharp/fslang-design/issues/712.

**Comment by nojaf** [comment]
This is available in the 5.1 series.
